osrf/vrx
# [Garden] Port WAMV-V xacro files and generation script

* Created a new `vrx_urdf` directory that consists of these packages:
  * `wamv_description` - Ported from `main` and updated xacro files to to use new gazebo systems
  * `wamv_gazebo` - Ported from `main` and updated xacro files (incl. `components` and `thruster_layout`) to use new gazebo systems and sensors
  * `vrx_gazebo` - ported the wam-v generation python and launch scripts to ROS2
* Updated the launch script in `vrx_gz` to support loading a urdf file of the wam-v
  * i.e. added `urdf` arg to `competition.launch.py`
  * The launch scripts convert urdf -> sdf before spawning the wam-v into simulation
* Gz<->ROS2 bridges for these new components / sensors should be working
  * `python-sdformat` is used to parse the generated sdf string in order to determine what sensors are attached to the wam-v and then launch the corresponding bridges. Note if you built sdformat from source, you will need to set the `PYTHONPATH`. See test instructions below.

To test:

1. Run the `generate_wamv` script to generate a urdf file from the provided example configuration yaml files.

    ```sh
    ros2 launch vrx_gazebo generate_wamv.launch.py component_yaml:=`ros2 pkg prefix vrx_gazebo`/share/vrx_gazebo/config/wamv_config/example_component_config.yaml thruster_yaml:=`ros2 pkg prefix vrx_gazebo`/share/vrx_gazebo/config/wamv_config/example_thruster_config.yaml wamv_target:=`pwd`/wamv_target.urdf wamv_locked:=False
    ```
    This should output a `wamv_target.urdf` in your current working directory.

1. Run the simulator with the generated wamv file

    a. If you installed sdformat from source, you will need to export your `PYTHONPATH`
        
       export PYTHONPATH=$PYTHONPATH:<path_to_colcon_ws>/install/lib/python 

    b. Launch sim

       ros2 launch vrx_gz competition.launch.py world:=sydney_regatta urdf:=`pwd`/wamv_target.urdf


![vrx_urdf](https://user-images.githubusercontent.com/4000684/194447462-3d352cc6-28d6-4a33-839c-027b9a463ebf.png)


```sh
ros2 topic list
...

/clock
/parameter_events
/rosout
/tf
/tf_static
/vrx/contacts
/vrx/task/info
/wamv/pose
/wamv/pose_static
/wamv/sensors/cameras/far_left_camera_sensor/camera_info
/wamv/sensors/cameras/far_left_camera_sensor/image_raw
/wamv/sensors/cameras/far_left_camera_sensor/optical/camera_info
/wamv/sensors/cameras/far_left_camera_sensor/optical/image_raw
/wamv/sensors/cameras/front_left_camera_sensor/camera_info
/wamv/sensors/cameras/front_left_camera_sensor/image_raw
/wamv/sensors/cameras/front_left_camera_sensor/optical/camera_info
/wamv/sensors/cameras/front_left_camera_sensor/optical/image_raw
/wamv/sensors/cameras/front_right_camera_sensor/camera_info
/wamv/sensors/cameras/front_right_camera_sensor/image_raw
/wamv/sensors/cameras/front_right_camera_sensor/optical/camera_info
/wamv/sensors/cameras/front_right_camera_sensor/optical/image_raw
/wamv/sensors/gps/gps/fix
/wamv/sensors/imu/imu/data
/wamv/sensors/lidars/lidar_wamv_sensor/points
/wamv/sensors/lidars/lidar_wamv_sensor/scan
/wamv/thrusters/left/pos
/wamv/thrusters/left/thrust
/wamv/thrusters/right/pos
/wamv/thrusters/right/thrust
```

If [`ground_truth_enabled`](https://github.com/osrf/vrx/blob/249141643aad288541bfbce8b92c47ab3c2e0483/vrx_urdf/wamv_gazebo/urdf/wamv_gazebo.urdf.xacro#L11) is set to `true`, there will be an extra `/wamv/sensors/position/ground_truth_odometry` topic

Things to note:

* The URDF -> SDF conversion process changes the collision names. The contact sensor currently hardcodes references to these collision names
* There is no `hector_gazebo_ros_imu` plugin in the new Gazebo. I used an imu sensor with parameters from mbzirc / subt
* There is no `hector_gazebo_ros_gps` plugin in the new Gazebo. The `navsat` sensor is used instead.
* There is no `gazebo_ros_p3d` plugin in the new Gazebo. The `odometry-publisher` system is used instead.
* The thruster systems have hard coded `xu`, `xuu` params for computing the max total thrust.
* The `pinger` and `ball shooter` plugins are commented out as they have not been ported yet
